Convert Marbles to SVG and rework diagrams

simonbasle opened the pull request:
Source for choice of fonts: http://www.ars-informatica.ca/article.php?article=59

Changes:

[doc] Rework marble diagrams into SVG
This commit is a large rework of the marble diagrams:
  - started from the omnigraffle original exported to SVG
  - large rework of diagrams with clearer and more consistent semantics
  - large naming rework to be more consistent with operator names
  - creation of missing marble diagrams
  - javadoc modifications to point to new diagram names
  - main javadocs no longer point to URLs but expect an embedded
  doc-files set up
  - SVG use Tahoma/Nimbus and Lucida Console/Courier New font families

[doc] Add "Contributing to SVG" README and task to copy SVG to doc-files

Said task is run as part of processResources, which gives a chance that new
users checking out the repo will see the quick documentation with diagrams
after first building the project.

simonbasle commented:
The diagrams are from #1377 by @thekalinga but the font family have been changed to not require any font download (and to be able to embed the SVG directly as is in the javadoc jar without mandatory text-to-path+optimization steps nor double committing of raw/optimized SVGs). Also the javadoc src changes have been reused from this PR with slight modifications and limiting the diff mostly to `<img>` rows.

simonbasle commented:
NB: now that the overall set up is finalized, we will need to validate the content of the diagrams themselves one by one vs the current release javadoc. Best way to do that is to produce a snapshot of the javadoc and visually compare it to the current released javadoc.

thekalinga commented:
@simonbasle This is not an issue per se. But thought it would help if you too know this. In some marble diagrams (`create` & `push` as far as I remember), svgs have java code that uses [Fira Code](https://github.com/tonsky/FiraCode) font that makes the code look much readable

This is the only place where I used font other than `IBM Plex Sans`. So incase you have not updated there aswell, please do so

simonbasle commented:
@thekalinga thanks for letting me know. Yeah I saw the `Fira Code` and replaced with "web safe fonts" `Lucida Console` (available by default on windows OS) with fallback to `Courier New` (available by default on most if not all OS)

simonbasle commented:
> @simonbasle One of the commit says, you converted text -> path.Not sure if I misunderstood the comment
> 
> If true, how will people be able to edit the svg in the future unless you have a text backup. Can you plz clarify
> 
> Commit Ref: [164f903](https://github.com/reactor/reactor-core/commit/164f9035f615274c66fc11a6b45e68d5db00be50) convention: timeout from text to path

@thekalinga fontawesome's clock glyph only, replaced it with a similar shape. people without fontawesome installed on their machine would not see the glyph, which is far more problematic.

simonbasle commented:
The only shape so far was the clock, which I replaced with a simplified hand drawn circle-and-two-lines. I don't want to have to look at licensing with Fontawesome Pro, we'll just get rid of it.

simonbasle commented:
Just in case this ever comes up, Fontawesome Free has the SVGs shapes covered by a Creative Commons By-SA licence apparently.

simonbasle commented:
Rough summary of changes:

<details>
<summary>First round of review started from the javadoc order</summary>

  - `combineLatestVarSources.svg`: missing arrow for transformation function, drop the extra params?
  - `combineLatestFixedSources.svg` & `combineLatestIterableSources`: drop the distinction with previous svg, missing lambda arrow too (seems to be linked to optimization?)
  - `concatWithValues.svg`: parenthesis problem (bad optimization?)
  - `concatAyncSources.svg` to be replaced by `concatAsyncSources.svg` (typo, part of URLs fixed)
  - `concatIterableSources` & `concatVarSources`: drop the parameter in name, share same diagram
  - `deferForFlux`: rework, still doesn't do defer justice
  - `error`/`errorWithSupplier`: drop the subscribe/onSubscribe. pb with light arrow
  - `errorWhenRequested`: drop the onSubscribe
  - `flux.svg`: ... a bit too on the left, "these are the item" shouldn't be this big/far up
  - `createForFlux.svg`: revert the opname and the inner diagram. use different visual for source to show not a `Flux`. Pb with light lambda arrow again.
  - `createWithOverflowStrategy`: same
  - `push`, `pushWithOverflowStrategy`, `createForFlux`, `createForMono`, `createWithOverflowStrategy`:
    all use a bezier curve shape to show the source is not a publisher (multithread vs singlethread)
  - `firstIterableSourcesForFlux`/`firstVarSourceForFlux`: drop params representation, combine into one, show late marbles in flux A
  - `fromForFlux`: put marbles together. find a way to show this is a Publisher not Flux, drop params
  - `fromArray`: drop params
  - `fromIterable`/`fromStream`: replace square brackets with parenthesis, drop params
  - `generate` and variants: find a new representation for the function/state, renamed withSink to Stateless
  - `mergeAsyncSources`: drop params
  - `mergeIterableSources`/`mergeFixedSources` drop params and combine into one
  - `mergeOrdered`/`mergeOrderedWith`: redo to match old diagram, drop params
  - `mergeSequentialAsyncSource`: put the top publisher inside box, drop params
  - replace `mergeSequentialIterableSources` with `mergeSequentialVarSources`, drop params replace with dots
  - `switchOnNext`: drop params, put colors in the Publisher<Publisher>
  - drop `switchOnNextWithPrefetch`
  - `usingWithCleanup*`: redone most of it, simplified parameters (color code), same for Mono
  - `usingWhen*` => turned into 4 entirely redone and separated diagrams for `Flux` and 2 for `Mono`
  - `zipTwoSourcesWithZipperForFlux.svg`: drop params other than lambda, drop request arrows, fix viewbox, fix output: A1 B2 C3 D4`
  - `zipFixedSourcesForFlux`: show 2 inputs, drop "tuple2" mention, fix output: A1, B2, C3, D4
  - `zipIterableSourcesForFlux`: fix lambda param, drop other params, fix output: A1 B2 C3 D4
  - `zipIterableSourcesAndPrefetchForFlux.svg` -> drop entirely and replace with `zipIterableSourcesForFlux.svg`
  - `zipVarSourcesWithZipperForFlux` -> drop entirely and replace with `zipIterableSourcesForFlux`
  - `zipAsyncSourcesForFlux`: this operator is weird. It is basically implicitely limited to 8 publishers in the source otherwise it falls apart :(. fix output A1...D4, fix params and lambda arrow.
  - `all`: fix lambda arrows,
  - `any`: swap true and false case for consistency with all, fix lamdba arrow
  - `blockFirst`/`blockFirstWithTimeout`: drop cogs, make return blocks start at subscribe, replace `request` with `subscribe`
  - similarly change `blockLast`/`blockLastWithTimeout`
  - `bufferWithMaxSize`, `bufferWithMaxSizeEqualsSkipSize`: replace 4 and 12 with original 2 and 6
  - `bufferWithMaxSizeLessThanSkipSize`: replace request numbers with original 3 and 9
  - `bufferWithMaxSizeGreaterThanSkipSize`: replace request numbers with original 3 and 7
  - `bufferWithBoundary`: make the blue one dropped and cancelling the source (remove from output)
  - `bufferWithTimespan`: complete before timeout: shift icon to right side
  - `bufferWithTimespanGreaterThanOpenBufferEvery` modify to reuse same source as `bufferWithTimespanLessThanOpenBufferEvery`, to stick to `buffertimeshiftover`
  - rework all of the above to better visually show the relationship between `bufferingTimespan` and `openBufferEvery` (eg. put each pair on the same line, remove the last openBufferEvery...)
  - `bufferUntilWithCutBefore`, `bufferUntil`, `bufferWhile`: missing lambda arrows
  - `bufferWhen`: remove most subscribe/cancel cruft. point to 3 flavors in javadoc text (dropped greens, overlapping)
  - `bufferWhenWithSupplier`: remove most subscribe/cancel cruft. point to 3 flavors in javadoc text (dropped greens, overlapping)
  - `cacheWithHistoryLimitForFlux`: removed last pink element and reorganized arrows
  - `cacheWithTtlAndMaxLimitForFlux`: removed last pink element and displayed smaller timeouts in a more compact way
  - `castForFlux`,`castForMono`: missing pentagon shape in the operator box
  - `collect`: missing lambda arrows
  - `collectList`: go back to rounded rectangle like the otherwise
  - `collectMapWithKeyExtractor`,  `collectMapWithKeyAndValueExtractors` remove `Map` mention, missing lambda arrow
  - `collectMultiMapWithKeyExtractor`, `collectMultiMapWithKeyAndValueExtractors`: remove `Map` mention, replace bracket with parenthesis, missing lambda arrow
  - `collectSortedList`, `collectSortedListWithComparator`: remove brackets/commas/weird rectangle in output, just rounded rectangle
  - `concatMap`: missing lambda arrow :(
  - `concatMapIterable`: have input element arrow stop at box, use parenthesis
  - `concatMapWithPrefetch`: drop and replace with `concatMap`
  - `defaultIfEmpty`: parenthesis misaligned
  - `delayElements` delays should be shifted a bit, all equal in the output, blue input marble shifted left a bit
  - Redid compose/transform marbles
</details>

<details>
<summary>Second round: remainder of marbles in alphabetical order</summary>

All the diagrams below have been verified with the optimization:

  - `autoConnect` + 2 variants, redone
  - `refCount` + 2 variants, redone
  - `block`, `blockWithTimeout`: reuse blocking convention
  - `blockOptional`, `blockOptionalWithTimeout`: reuse blocking convention
  - `bufferTimeoutWithMaxSizeAndTimespan`: timeouts on single line
  - `cacheForFlux`: simplify 2 marbles (`cacheForMono` ok)
  - `cacheWithTtlForFlux`, `cacheWithTtlForMono`: show timeouts under separator at bottom of op block
  - `collectWithCollector`: mention parameter is a `Collector` (no good idea for representation)
  - `concatWithForFlux`, `concatWithForMono`: remove parameter
  - `deferForMono` follow same change as `deferForFlux`
  - `delay`: show result is 0L, remove orange from signal arrows
  - `delaySequence`: only show first (most significant) delay
  - `delaySubscriptionForFlux`, `delaySubscriptionForMono`: realign timeout parameter
  - `delaySubscriptionWithPublisherForFlux`: show companion publisher (new convention)
  - `delaySubscriptionWithPublisherForMono`: show companion publisher (new convention)
  - `delayUntilForFlux`: better show the concatMap nature of the delays, adjust `delayUntilForMono`
  - delete `directProcessor.svg`
  - `dematerializeForFlux`: remove icon and realign
  - `dematerializeForMono`: make the source a `Mono<Signal>` (split in 3), remove icon
  - `directProcessor.svg`: remove, will need more work to represent processors
  - `distinctUntilChangedWithKey`: remove color from label and reintroduce lambda arrow
  - `distinctWithKey`: realign and lambda arrow
  - `do*`: simpler side effect convention, fix "doAfter"
  - removed `emitterProcessor.svg`
  - fixed elapsed to remove Tuple2 mention
  - deleted few unused files
  - `firstForMono` (removed border color)
  - `fromSupplier`: lambda
  - `fromRunnable`, `fromCallable`: show subscribe, execution block, lambda arrow
  - `fromFuture`: new convention (Future), show immediate start before subscription
  - `fromFutureSupplier`: new convention (Future), show deferred start at subscription
  - `filterForFlux`/`filterForMono`: fix lamdba arrow
  - `filterWhenForFlux`/`filterWhenForMono`: 4 cases: mono true, mono false, empty, cancelled flux
  - `flatMapMany`: lambda arrows
  - `flatMapForFlux`/`flatMapForMono`: missing lamda arrow
  - `flatMapIterableForFlux`/`flatMapIterableForMono`: missing arrow, put the iterable under the source element arrow
  - `flatMapManyWithMapperOnTerminalEvents`, `flatMapWithMapperOnTerminalEventsForFlux`: vastly simplify (no inner details), black color for completion publisher
  - `flatMapWithConcurrency`, `flatMapWithConcurrencyAndPrefetch`: lambda arrow
  - `flatMapSequential`, `flatMapSequentialWithConcurrency`, `flatMapSequentialWithConcurrencyAndPrefetch`: lambda arrow
  - `groupJoin`: realign inners and result with original diagram (1A, 2A, 3B, 4B, 4C)
  - `groupByWithKeyMapper`,`groupByWithKeyMapperAndValueMapper`: avoid the rotated flux style (mini flux + horizontal style), use "noBorder"
  - fix b-e marbles with clock fontawesome to shape
  - `hasElementForFlux`, `ignoreElementForMono`, `ignoreElementsForFlux`: black arrows, realign a bit
  - `ignoreElementsForMono` use static convention: no param, use classic input flux, remove gray arrows
  - `interval`: replace fontawesome clock, elipsis as text
  - `join`, redo like `groupJoin`, realign with original marble, fixed output
  - `just*`, `last*`: minor fixups
  - `logForFlux`, `logForMono`: use a logger representation similar to side effects, yet reminiscent of a console output
  - `monoProcessor`: deleting processor diagrams for now
  - `mapForFlux`, `mapForMono`: lambda arrow
  - `merge*`: small alignments. async parameter versions have the Flux<Flux> displayed
  - `next`, `never`
  - `onBackpressureBuffer`: show buffer as rounded rectangle under op name
  - `onBackpressureBufferWithMaxSize`: actually show max size and dropping when 3 extra elements come in, rounded rectangles
  - `onBackpressureDrop`: gray arrow to black
  - `onBackpressureDropWithConsumer`: gray arrow to black, show consumer as {} block under opname
  - `onBackpressureLatest`: use same rounded rectangle convention as onBackpressureBuffer
  - `onErrorContinue*` redid a simplified version. show the consumer in upstream operator, show upstream configuration arrow
  - `onErrorMap*`: redraw crosses
  - `onErrorResume*`: redraw crosses, lambda arrows
  - `onErrorReturn`: redraw crosses, realign text
  - `orForFlux`/`orForMono`: single example, show cancelled elements
  - parallel, publish, publishOn and subscribeOn marbles
  - `publishNext`: show cancelled element
  - `publishOn`/`subscribeOn`: adjust where the orange thread starts, blue subscribe
  - `parallel`: make the output a double Flux (rails) instead of rotated arrows
  - `range`: rounder shape for last element
  - `reduce`/`reduceWith`/`reduceWithSameReturnType`: missing lambda arrow
  - `repeatForFlux`/`repeatForMono`: make resubscribe arrow blue and don't show requests
  - `repeatWhenForFlux`/`repeatWhenForMono`: use companion convention, don't show too much variants
  - `repeatWithPredicate`..`repeatWithAttempts`..`repeatWithAttemptsAndPredicate`: adapt to use same conventions
  - deleted `replayProcessor*` diagrams
  - `replay`, `replayWithTtl`, `replayWithHistory`, `replayWithHistoryAndTtl`: add notion of connect(), replace fontawesome clock
  - `retryBackoffForFlux`: show gray "duration" backoffs, remove most cruft text and show x2/x4/jitter visually
  - `retryBackoffForMono`: same, finish with value
  - delete too specific variants of `retryBackoff*`
  - `sample*`: replace clock font with shape, companions publishers with convention
  - `sampleTimeoutWithThrottlerFactoryAndMaxConcurrency`: for now reproduce the same marble as `sampleTimeoutWithThrottlerFactory`, added a FIXME to reevaluate diagram and javadoc
  - `scan*`: missing lambda arrows, adjust curvy arrows being hidden being operator text
  - `single*`: small alignment adjustments
  - `skip`, `skipLast`: distinguish first request on subscribe vs further request, change to bigger marbles
  - `skipUntil`, `skipUntilOther`: bigger marbles, companion convention
  - `skipWhile`: bigger marbles, revert boolean, remove requests
  - `skipWithTimespan`: bigger marbles, show subscribe and shift timeout, replace clock font with shape
  - `startWith*`: remove params, use parenthesis convention for iterable, big marbles
  - `subscribe*`: use side effect conventions, no out marble
  - `subscribeOn*`: fix marble fill, small alignments
  - `switchIfEmpty*`, `switchMap`, `switchOnNext`: minor adjustments (margin, fill)
  - `switchMapWithPrefetch`: remove that one
  - `take*`: bigger marbles
  - `thenForFlux`: show 1 marble in the error case
  - `thenReturn`: realign parameter
  - `timeoutForFlux`, `timeoutForMono`: clock font to shape, simplify to main case.
  - `timeout*`: simplified names, clock font to shape, simplify to main case, removed distinction Mono/Flux for `timeoutPublisher`
  - `timestampForMono`/`timestampForFlux`: remove tuple2, show a black clock on the time axis
  - removed `toProcessor` and `topicProcessor` marbles
  - removed `whenIterable`/`whenDelayErrorIterable`, a bit too redundant
  - `windowUntil`: fixed green boolean, marble arrow, bigger source marbles, alignment
  - `windowUntilWithCutBefore`: same changes, only show cutBeforeTrue case
  - `windowWhile`: same changes, fix diagram to show that windowWhile drops elements
  - `windowWhen`: using open/close convention, realign and fix marbles, using big ones as source
  - `windowWithMaxSize*`: reuse same request numbers as before
  - `windowWithBoundary`: use boundary convention same as with bufferWithBoundary
  - `windowTimeoutWith*` renamed to `windowTimeout`, put timeouts on same line with cancelled timeout convention
  - `windowTimespan*`: fixed clock, position and size of timeout/timespan, param names
  - `withLatestFrom`: fixed wrong output
  - various additional Mono `zip*` adjustments
  - `zipWithIterableForFlux`,`zipWithIterableUsingZipperForFlux`: remove tuple2 mention, use curly brackets, rework cancel representation
  - removed `zipWithOtherUsingPrefetchForFlux` (`zipWithOtherForFlux`)
  - removed `zipWithOtherUsingPrefetchAndZipperForFlux` (`zipWithOtherUsingZipperForFlux`)
  - `zipWith*`: fix output (A1, B2, C3, D4)
</details>

thekalinga commented:
@simonbasle I have a bunch of comments & suggestions on the new svgs (+ conventions). Will share them here today/tomorrow

thekalinga commented:
@simonbasle 

I have attached a bunch of questions/suggestions on the latest version of marble diagrams. Please let me know if you find any of my suggestions/queries

Thanks

[Comments - Questions - Suggestions on
Marble diagrams.pdf](https://github.com/reactor/reactor-core/files/2621753/Comments.-.Questions.-.Suggestions.on.Marble.diagrams.pdf)

thekalinga commented:
@simonbasle 

After reading my suggestions one more time, I realised some of my suggestions cannot be understand with my cryptic text. So, I updated the suggestions to include more context

Also, I have created some mockups for the suggestions I make

You can find updated version of this suggestion document & mockups for new suggestions in the attachment

Please let me know if you have any suggestions

[Comments - Questions - Suggestions on Marble diagrams - v2.pdf](https://github.com/reactor/reactor-core/files/2623979/Comments.-.Questions.-.Suggestions.on.Marble.diagrams.-.v2.pdf)
[suggestion-mocks.tar.gz](https://github.com/reactor/reactor-core/files/2623980/suggestion-mocks.tar.gz)

simonbasle commented:
@thekalinga I will work on the suggestions below that have a checkbox, other ones are my answers to your comments. I mentioned you below explicitly for stuff I'm not sure how to easily fix in inkscape.

### common suggestions
I don't agree / don't think it is worth the wide change for most of these:
 - `Throwable` class predicate as outline would make the difference between exception types even less visible
 - operator text & operator misaligned: do you have an example? (screenshot) I don't think it is super critical (unless the operator text is very small and very noticeably shifted left/right)
 - "replace harsh black color": The gray color is what made the lambda arrow disappear after `svgo` in almost 100% of your original diagrams, so no way.
 - "follow xxx operator style in xxx": don't want to touch the old operator marbles for concat/merge/flatmap/etc...
 - arrow for exception does not start just below exception: don't want to spend any more time on marbles than necessary, don't think this one is very important

The ones I think are worth it:
 - [ ] using boolean icon instead of "true"/"false" boxes in downstreams
 - [ ] buffer boxes corners are uneven: inkscape makes it so hard to keep the same corner curve while increasing/decreasing the box's width 👿. any suggestion @thekalinga ?

### Conventions comments
Where I don't agree:
 - `create`: not touching this one, might consider 2 close colors but that's it
 - `sink`: no better idea to represent the sink, so status quo
 - `Future` not aligned: screenshot? there is no particular alignment in `conventions.svg`, it's just a collection of unrelated representations to reuse in diagrams...
 - `Flux` vs `Publisher`, yes it is required
 - companion publisher subscription: meh, maybe to be fixed on a case by case basis later if people express they got confused. IIRC most actual diagrams show a more correct subscription pattern, but in any case it can't be perfect
 - side effect should be shown inside operator: need to think more about this, but I think the fact there's no arrow is pretty clear and gives `subscribe` a very distinctive look.
 - "can the upstream side effects": what are you talking about? The 3 side effect styles are: the ones to use in parameter blocks/functions, the ones to represent effects happening _before_ the signal is propagated downstream and the ones to represent effects happening _after_ the signal is propagated downstream. The last 2 are mirrored, as shown in the example below the styles.

The ones I agree with:
 - [ ] color of function/bifunction/etc...: all should be hard black
 - [ ] downstream values "true/false": use boolean flag icons
 - [ ] **some colors have opacity/transparency**: this is hard to spot in inkscape, but should definitely be fixed. any suggestion to easily find affected diagrams @thekalinga ?
 - [ ] blocking side effect can actually be removed from convention, not used

### Individual diagrams comments:
 - `autoconnect`: the ConnectableFlux cases are more complicated: we need to show there is an upstream (as in "the operators are not static"), but the `ConnectableFlux` hides it from downstream, delaying subscription. The only alternative I can think of is for the arrow to traverse the operator block, but that would look ugly with the operator name in the middle
 - [ ] `bufferWhen`: can't visually spot a difference... however looks like the blue source marble is missing.
 - [ ] `cache*`: I'll change all the cache variants to start the arrow at request point
 - [ ] `transform`/`compose`: I'll try to make it clearer that transform happens immediately
 - [ ] `count` crooked outline. original cut corner outline, must not be that compatible with `svgo`, but yeah I can change it to plain rectangle
 - [ ] `deferForMono`: will fix text gap
 - [ ] `delay`: will fix text gap and replace output crooked boxes with rectangles
 - [ ] `doOnErrorWith*`: will use more contrasted color instead of orange
 - [ ] `doOnSuccess`: will fix gap
 - [ ] `flatMapMany...`: not touching these
 - `fromCallable`/`fromRunnable`: but we **do** need to show that if the runnable blocks, it blocks/delays the completion of the operator. so 100% intentional.
 - `fromFuture`: can't spot the color difference
 - [ ] `generateStateless`: will rework the clock icon
 - [ ] `generateWithCleanup`: will move the cleanup inside the operator box (see `onBackpressureDropWithConsumer` for rationale)
 - [ ] `log*`: will make log container square
 - [ ] `collectSortedListWithComparator`: will use same comparator as in `mergeOrdered`
 - `onBackpressureDropWithConsumer`: drop consumer is more an "inside" concern of the operator rather than something that can contain business logic, unlike side effects
 - [ ] `onErrorContinue`: that's the trouble trying to represent a pretty complex operator in a simple enough diagram. will try to show more that the operator affects operators upstream of it. inner box is internal execution of cleanup consumer.
 - [ ] `publish`/`replay`: yellow dot is when a downstream signal triggers the connection/disconnect and subsequent upstream signals. icon is when the user manually triggers. Can remove the icon and make arrow purple to show external API though.
 - [ ] `repeatWithAttemptsForFlux`,`repeatWithAttemptsForMono`,`repeatForFlux`,`repeatForMono`: will get rid of the first arrowhead in the resubscribe arrows
 - [ ] `repeatWhen*`: can't replace with same style as in retryWhen because the transformed companion is basically a filter. will show `!` in transformed companion elements.
 - [ ] `retryWithAttemptsForFlux`,`retryWithAttemptsForMono`,`retryForFlux`,`retryForMono`: will get rid of the first arrowhead in the resubscribe arrows
 - [ ] `retryWhen*`: show `!` in transformed companion elements
 - `retryBackoff*`: no, green color is for timeouts or ttl. here we _introduce_ a delay.
 - [ ] `sampleAtRegularInterval`: correct, will align orange downstream with complete and cut short the last timeout
 - [ ] `sampleFirstAtRegularInterval`: will add a last short timeout for completion
 - `subscribe*`: the absence of downstream arrow makes it pretty clear this is not a Flux, and the "effect outside the box" give the `subscribe` marbles a distinctive look so I think I'll keep it that way.
 - [ ] `timeoutFallBack*`/`timeoutPublisherAndFallback*`: will remove color of parameter flux
 - [ ] `usingWhen*`: top rectangle is state of the resource. flux with green square is "what to do in case of success". red square is "what to do in case of failure". When former one fails, it results in the downstream flux failing (even though the derived flux didn't fail, only the "commit" flux failed). Will change the orange color of the resource to blue to show it's a good state, not a "warning" or "error" as the color currently suggests. any other deeper change will probably need a separate issue.
 - [ ] `windowWhen`/`bufferWhen`: will replace the icons entirely, not happy with it. will probably use curly braces text.

simonbasle commented:
After having passed way too much time on this PR, I won't go over all the diagrams to fix the colors / transparency / etc... I've done it for `conventions.svg` (which now includes an explicit color palette) and tried to best effort fix the palette in the diagrams I otherwise edited per your review @thekalinga, and I feel that is good enough for now.

We could maybe add a sentence in the README to the effect of:
> if you notice color differences from the official palette in a preexisting diagram you are editing, that is probably legacy hiccups and you mighty also fix the colors. For new diagrams, try to start from the elements in `conventions.svg` to avoid this issue.

wdyt?

thekalinga commented:
@simonbasle Sorry for the delay in responding. Shifted my house this weekend. Was fully occupied the last few days. Still have limited connectivity (working with my mobile internet)

Some of the items I mentioned like upstream sideeffect are removed from `conventions.svg` in your latest  commit. Also, wherever we disagree, I defer to your judgement on intuitiveness. So I'll keep my response focused on these latest version of the marbles

> companion publisher subscription: meh, maybe to be fixed on a case by case basis later if people express they got confused. IIRC most actual diagrams show a more correct subscription pattern, but in any case it can't be perfect

Concur with you. But I have an observation to make. These marbles will be referred by a vastly disproportionate number of developers when understanding the flow. My view is that, its better to represent them as accurately as possible (- my nitpicks). So any effort we put on these will have vastly disproportionate amount of usefulness to users of the library. But am also happy with the state of most of the marbles in the latest version

> can the upstream side effects
This is removed from the latest version of `conventions.svg`. We can ignore this now

>  some colors have opacity/transparency: this is hard to spot in inkscape, but should definitely be fixed. any suggestion to easily find affected diagrams

Not sure if the following two regexs would find all the non-1 opacities, but one of them yielded matches inside `ofTypeFor*.svg` which needs to be corrected

```
// regex to catch one version of opacity problems
opacity\s*:\s*[^1]+\s*;?
// regex variation to catch equals version of opacity problems
opacity\s*=\s*"[^1]+?"
```

> autoconnect: the ConnectableFlux cases are more complicated: we need to show there is an upstream (as in "the operators are not static"), but the ConnectableFlux hides it from downstream, delaying subscription. The only alternative I can think of is for the arrow to traverse the operator block, but that would look ugly with the operator name in the middle

The latest version is even more confusing as we have one stream above the operator & and another one in the middle of the operator

>  transform/compose: I'll try to make it clearer that transform happens immediately

The latest version is better than the previous version. But I am not sure about the thick arrow. Also, does compose/transform operator participate in downstream events (lets say logging)? If it does participate, the latest version seems superior to the version I suggested previously, else can you look at the version I previously proposed and see whether that represents the flow more accurately

>  usingWhen*: 4top rectangle is state of the resource. flux with green square is "what to do in case of success". red square is "what to do in case of failure". When former one fails, it results in the downstream flux failing (even though the derived flux didn't fail, only the "commit" flux failed). Will change the orange color of the resource to blue to show it's a good state, not a "warning" or "error" as the color currently suggests. any other deeper change will probably need a separate issue.

The confusion is mainly because there are downward arrows which we use for downstream elements is being reused (arrow from resource to stream). Please correct me if I am wrong, the resource by itself does not produce anything rather the operator applies the function to get the reference to a new stream by passing in the resource, to which it later subscribes. So the downward arrows originating from resource (both at the start & before the begining of the success/end stream) should be removed

### Alignment issues

![uneven spacing in operator boxes](https://user-images.githubusercontent.com/452518/49343683-42bbd980-f694-11e8-900d-0f80d44a87e9.png)

![odd gaps between operator text arguments](https://user-images.githubusercontent.com/452518/49343755-0b99f800-f695-11e8-8a5a-03ea66172248.png)

Almost every operator I see in the latest version have this problem of having a lot of space between operator name & the actual arguments

Strangely I am seeing this only in chrome, but firefox is showing **almost** evenly spaced items

Besides that the operator + arguments are not center aligned along y -axis w.r.t the box surrounding it. This animation below shows how to fix it

[How to align gif](https://im5.ezgif.com/tmp/ezgif-5-564c6feb62ab.gif)

Also, `flux.svg` & `mono.svg` are shown completely distorted in my inkscape but not in browsers. Inkscape magic I guess :D

![flux](https://user-images.githubusercontent.com/452518/49344058-7c8edf00-f698-11e8-99fd-7559325ab495.png)

![conventions - future evenly spaced](https://user-images.githubusercontent.com/452518/49343661-b7dadf00-f693-11e8-9eb4-4abda6472e78.png)

In the above diagram, I have corrected few alignment issues and spaced them evenly just for the record

![alignment issue](https://user-images.githubusercontent.com/452518/49343737-cece0100-f694-11e8-8f90-21e5a755ea2e.png)

`autoConnect` alignment issue. The top green marble is not in alignment with the middle one. Not sure if its intentional

![fromfuture - color difference](https://user-images.githubusercontent.com/452518/49343749-fa50eb80-f694-11e8-910e-81e2a7f3e6c7.png)

`fromFuture` color difference. Black on the left, lighter black on the right

### Good to have

1. Can we have this for Signal (RS level method call) along the line (vs) every other call with a dot as starting point. so that it is vastly clear to users which one is which

#### Previous version

![signal vs method call old](https://user-images.githubusercontent.com/452518/49343654-937f0280-f693-11e8-892b-21a58def4c93.png)

#### Suggestion

![signal vs regular call new](https://user-images.githubusercontent.com/452518/49343637-574ba200-f693-11e8-9547-fbb4b44630e3.png)

simonbasle commented:
I don't have the alignment issues in inkscape, and neither in chrome/firefox. eg. for Future:
![image](https://user-images.githubusercontent.com/6986166/49363966-5d696f00-f6e3-11e8-83e4-33cf03fdb644.png)

thekalinga commented:
@simonbasle

Inkscape version 0.92.3

simonbasle commented:
> Inkscape version 0.92.3

Well, shit...
![image](https://user-images.githubusercontent.com/6986166/49367528-a1ad3d00-f6ec-11e8-8675-cff024b50815.png)

I recognize that Inkscape is a lower barrier to entry to contributions that a proprietary software (by far), but the experience with it on OSX has often been disappointing and frustrating. X11/XQuartz is of course a major factor there...

Unfortunately, I can't find an alternative SVG editing software that manages to correctly open the most basic of our marbles, since all of those I have tested removed arrowhead on first opening a file 🤕

simonbasle commented:
@thekalinga do you prefer the alternative version for `usingWhen` I just pushed? see https://github.com/reactor/reactor-core/pull/1390/commits/f04990530d04908767ad843d0a5ae7d992a64d0d

thekalinga commented:
@simonbasle They are capturing the details more or less. However, I am not sure whether extending the item to resource it causes confusion (or) not

Here are some minor tweaking I did just to keep the method calls look the same everywhere

## Suggestion 1

### Present

![userwhencurrent](https://user-images.githubusercontent.com/452518/49371583-fc12c180-f71d-11e8-859c-d32503bbb11e.png)

### Suggestion

![usingwhensuggestion](https://user-images.githubusercontent.com/452518/49371601-06cd5680-f71e-11e8-86f9-302a7bed41c9.png)

## Suggestion 2

### Present

![tofutureexisting](https://user-images.githubusercontent.com/452518/49371741-7b07fa00-f71e-11e8-8862-a665eba31e14.png)

### Suggestion

![tofuturesuggestion](https://user-images.githubusercontent.com/452518/49371748-822f0800-f71e-11e8-8a5a-7d034177da97.png)

## Suggestion 3

### Present

![createforflux](https://user-images.githubusercontent.com/452518/49371772-94a94180-f71e-11e8-874e-9c2acbc33793.png)

### Suggestion

![suggesion](https://user-images.githubusercontent.com/452518/49371778-98d55f00-f71e-11e8-8961-1f7425dedc1a.png)

These are only minor tweaks. If they cause less confusion (like `create` invoking `addListener` instead of `Sink`[current version feels like `sink` is actually making the call]) we can incorporate them, else we can live with what we have for now

thekalinga commented:
@simonbasle 

> Unfortunately, I can't find an alternative SVG editing software that manages to correctly open the most basic of our marbles, since all of those I have tested removed arrowhead on first opening a file face_with_head_bandage

Oh ok.. Another gotcha.. Can we stick to the lower version for now then till Inkscape starts supporting mac?

simonbasle commented:
@thekalinga integrated your latest suggestions. didn't integrate the `usingWhen` subscribe arrow as I'd prefer to use these for **re**-subscriptions (like in repeat/retry) only

thekalinga commented:
@simonbasle 

Minor observations on the latest push

1. `usingWhenSuccessForFlux.svg` - For this, downward arrow is missing for the complete element of the resource stream (when compared to `usingWhenSuccessForMono.svg`)
2. Can the `rollback/commitThrowing` methods also be vertically aligned when it comes to cleanup on success/error for resources
3. `pushWithOverflowStrategy.svg` - Can the ping arrowhead be extended slightly downwards inside the sink so that it looks like the yellow item, i.e this item will be temporarily stored inside sink till downstream requests more items
4. `usingWhenCleanupErrorForFlux.svg` - Can the downward arrow be removed from the resource to exception on cleanup

simonbasle commented:
@thekalinga done!

thekalinga commented:
@simonbasle 

`usingWhenSuccessForFlux.svg` operator box seems to be not aligned with the content

![usingwhen](https://user-images.githubusercontent.com/452518/49383633-32146d80-f73f-11e8-8b7e-f80986d40e2d.png)

Finally, I will do one round of verification of each operator from start to finish and share any final observations lil later today

simonbasle commented:
@thekalinga fixed and rebased to combine today's usingWhen changes into a single commit

thekalinga commented:
@simonbasle 

Here are my latest (and hopefully last on marbles) set of observations on the recent set of marbles. Some of these are trvial, but some are not. Can you please go over them and see if any of these changes are worth doing

1. `bufferWithBoundary.svg` - Boundary item is not center aligned inside the circle
2. `cacheWithTtlAndMaxLimitForFlux.svg` - Second subscriber only got 2 items, not sure if the `maxLimit` parameter to cache be `2` instead of `3`
3. `cancelOnForMono.svg`, `subscribeOnForFlux.svg`, `subscribeOnForMono.svg`, `` - It feels a bit odd that the `downstream -> operator cancel signal` is in blue color, where as `operator -> upstream signal` is in orange color (+ downstream elements are in black color). Please note that I understand the convention we are following but just trying to think it from the point of end user who encounters this to understand what is going on. May be we should just use black color for `downstream -> operator cancel` signal as an exception to just this operator (+ its siblings)
4. `concatAsyncSources.svg`, `mergeAsyncSources.svg`, `mergeSequentialAsyncSources.svg`, `switchOnNext.svg` - Can the arrow from upstream be in thick accent (like `groupByWithKeyMapper.svg`) rather than what we have now, as it refers to a stream instead of a regular element
5. `createWithOverflowStrategy.svg` - This also requires the arrowhead of pink element to be inside sink
6. `delay.svg` - Should we use orange color for the downward pointing arrows aswell as they are emitted on a different thread? (just like `interval.svg`)
7. `delayElement.svg`, `delayElements.svg`, `delaySequence.svg`, `delaySubscriptionForFlux.svg`, `delaySubscriptionForMono.svg`, `delaySubscriptionWithPublisherForFlux.svg`, `delaySubscriptionWithPublisherForMono.svg` - Above is applicable to this too
8. `delayUntilForMono.svg` - Lamdba arrow missing
9. `doOnErrorWithClassPredicateForFlux.svg`, `doOnErrorWithPredicateForFlux.svg` - Can we use yellow color for the second marble instead of orange like we did for `doOnTerminateForFlux.svg`
10. `flatMapManyWithMappersOnTerminalEvents.svg` & `flatMapWithMappersOnTerminalEventsForFlux.svg` - These two are not as clear as they can be when compared to all other `flatMap*.svg`s. Its better to keep this also consistent with all other `flatMap*.svg`
11. `fromFutureSupplier.svg` - Can we have `external call` called `get` instead of the arrow
12. `generateStateless.svg` - There is a missing closing brace
13. `groupJoin.svg`, `join.svg` - The first argument for this is missing (the other publisher), not sure if its intentional
14. `logForFlux.svg`, `logForMono.svg` - Not sure if we intentionally did not represent `subscribe`. Don't remember exactly but is `subscribe` not logged if we enable logging?. If logged can we represent this aswell in thi diagram
15. `mergeOrderedNaturalOrder.svg` - Can we remove braces at the end of the operator
16. `onErrorContinue.svg` - Is the consumer only triggered only on next item (or) just on error. If just on error, can we move the blue marble & completion signal to the rightmost so that its more clear to the user
17. `onErrorMapWithPredicateForFlux.svg`, `onErrorMapWithPredicateForMono.svg` - We have outline version here for exception for some reason
18. `publishNext.svg` - Can these arrow for subsequent subscribe also start from the request point instead of the elements
19. `range.svg`, `repeatWhenForFlux.svg`, `repeatWhenForMono.svg`,  - Can we increasing the size of marbles slightly so that the text fits the marble in a single line
20. `refCount.svg`, `refCountWithMinSubscribers.svg` - Top level items & middle level items are a bit confusing as they are not aligned. I assume you meant to say these are common elements expected between two scenarios
21. `repeatWithAttemptsAndPredicateForFlux.svg`, `repeatWithAttemptsAndPredicateForMono.svg`, `retryWithAttemptsAndPredicateForFlux.svg`, `retryWithAttemptsAndPredicateForMono.svg` - Can we change arg to `3/4` instead of `2` so that it ois explicit to the user that the maxRetries did not trigger the completion in the top scenario, rather its just boolean (tho its obvious to us, new users might get confused about this. If we make it 3/4, there is no ambiguity)
22. `sampleTimeoutWithThrottlerFactory.svg`, `sampleTimeoutWithThrottlerFactoryAndMaxConcurrency.svg` - Both have the same diagrams
23. `skip.svg`, `skipLast.svg` - Can we move `request(n)` to be in between green and yellow marble so that user gets a more accurate idea
24. `toIterable.svg`, `toStream.svg` - Can we specify whether the iteration start method is `hasNext`/`next`
25. `usingForFlux.svg`, `usingForMono.svg`, `` - Can we use dot notation & horizontal text for `close`
26. `windowWithMaxSizeGreaterThanSkipSize.svg` - Can you increase the size of the font for small text slightly. They are almost invisible at current font size

simonbasle commented:
@thekalinga fixed, and/or commented below:

1. (probably due to precision trimming in the optimization? I realigned but it barely moved it)
2. the second subscriber demonstrates the TTL: blue marble has expired before it subscribed
3. I was torn on this and decided to follow the convention. I don't expect much confusion, but if somebody voices it, we'll change it back later.
7. did this for `delay` and thanks to svgo inline styling and inkscape crappy UX with markers it was a massive pain (ended up editing the marker xml manually). won't go through that trouble again, let's say we did it for _source_ operators that _generate_ a delay.
10. cannot be consistent as they are not exactly the same and mirror what we had originally. it is therefore out of scope for this PR.
13. intentional, since we decided not to show input publishers in the parameters block _most of the time_ unless it is adding extra value, to focus on functions for instance
14. nope, only `onSubscribe` is visible to log.
16. `onErrorContinue.svg` the consumer is triggered within `onNext` if there is an exception, to swallow said exception and request one more.
17. outline throwable because we've used the color to represent the `class`. outline as a way to say "this is a `Predicate<Throwable>`, not a `Predicate<RedException>` or `Predicate<PurpleException>`
19. done for `range`
20. top element is the source of the `ConnectableFlux`, inner element is the representation of _when_ the source is connected to.
24. changed all 4 (`toIterable`x2 and `toStream`x2) to make that clearer
26. `windowWithMaxSizeGreaterThanSkipSize.svg` font size: done for all variants, plus buffer variants, putting on a single line at bottom left of operator box (made box bigger and operator name shifted up)

thekalinga commented:
@simonbasle Went over the changed files & they look almost final

Reg (16), if the meaning remains unchanged, instead of what we have now, can we have this as marble diagram (blue marble & complete signals moved right)

![onerrorcontinue](https://user-images.githubusercontent.com/452518/49871497-70471680-fe3c-11e8-8b09-05368db4b91f.png)

This picture makes it clear that `onErrorContinue`'s consumer invocation has nothing to do with when the future events happen (unlike current version)

### Something for Future

I have this mental model w.r.t class vs object

Class: Defines the structure/skeleton of an object (be it a regular class/throwable)
Object/Throwable: an instantiated version of Class

Based on these guidelines, I have created this convention that has simple rules that can be followed

![class vs object](https://user-images.githubusercontent.com/452518/49871857-61149880-fe3d-11e8-85a3-0772396eb256.png)

In our marbles we dont distinguish between class & object. I believe this distinction can easily achieved with the above simple rules. Can you check if this can be accommodated into marbles in the future

simonbasle commented:
- `onErrorContinue`: ok will move the blue elements to the right.
 - class vs instance: the number of diagrams in which the distinction is super relevant is very limited (a few diagrams with the notion of class predicate), so adding a convention with as many variant would add too much noise in my opinion

thekalinga commented:
@simonbasle We are good to go then :+1:

simonbasle commented:
Rendering of marbles inside jar tested in Eclipse (might be subpar, some arrowheads look weird, but let's see if people complain) and IntelliJ IDEA 2018.3.1 (rendering size bug has been fixed in this version)

simonbasle commented:
@bsideup so far no Flux combining operator has a `*DelayError` variant of the diagram. I don't think it is super critical, but if it comes up a couple times more we can always add them.
The `get()` is from the `Supplier`. I added `Supplier#get()` to make it clearer.

simonbasle commented:
Between xmas and new year, 'tis time to merge this PR 🎅
